net/captivedetection/captivedetection.go

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package captivedetection provides a way to detect if the system is connected to a network that has
+// a captive portal. It does this by making HTTP requests to known captive portal detection endpoints
+// and checking if the HTTP responses indicate that a captive portal might be present.
+package captivedetection
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"runtime"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"tailscale.com/net/netmon"
+	"tailscale.com/tailcfg"
+	"tailscale.com/types/logger"
+)
+
+// Detective is a type used to detect if the system is behind a captive portal.
+type Detector struct {
+	// mu is a mutex used to protect httpClient and currIfIndex from concurrent access.
```

Comment:
So httpClient is mutable? If so, you also need to hold the mutex while using it. But that'll get gross probably.

I'd prefer to make the httpClient not be mutable. You could instead make a fixed http.Client+http.Transport+Dialer where the DialContext func is an unexported method on Detector and that method takes the mutex to read the currIfIndex.